File: sol/solution/0001-0100/0017/Solution.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
	if (!digits || digits.length === 0) return [];
	const map = {
		'2': 'abc',
		'3': 'def',
		'4': 'ghi',
		'5': 'jkl',
		'6': 'mno',
		'7': 'pqrs',
		'8': 'tuv',
		'9': 'wxyz'
	};
	const result = [];
	function backtrack(index, path) {
		if (index === digits.length) {
			result.push(path);
			return;
		}
		for (const char of map[digits[index]]) {
			backtrack(index + 1, path + char);
		}
	}
	backtrack(0, '');
	return result;
};

File: sol/solution/0001-0100/0017/Solution.scala
object Solution {
    def letterCombinations(digits: String): List[String] = {
        val digitToLetters = Map(
            '2' -> "abc",
            '3' -> "def",
            '4' -> "ghi",
            '5' -> "jkl",
            '6' -> "mno",
            '7' -> "pqrs",
            '8' -> "tuv",
            '9' -> "wxyz"
        )
        if (digits.isEmpty) return List()
        digits.foldLeft(List("")) { (acc, digit) =>
            for {
                combination <- acc
                letter <- digitToLetters.getOrElse(digit, "")
            } yield combination + letter
        }
    }
}